AIFC - Enablement of AIFC development object for AIF content in abapgit

D047539 opened the pull request:
To enable shipment of AIF content via ABAP Git we implemented the new AIF serializer zcl_abapgit_object_aifc. For more information see issue #5195.

Please review the object of the pull request and merge them into the repository.

larshp commented:
thanks, can you help opening a PR to https://github.com/abapGit-tests/AIFC with a simple example of the AIFC object

we keep examples for testing and future reference

larshp reviewed (changes requested):
please open PR in https://github.com/abapGit-tests/AIFC with example files

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,673 @@
+    TRY.
+        lv_meth     = '/AIF/IF_ABAPGIT_AIFC_UTIL~COMPRESS_INTERFACE'.
+
+        ls_parameter-name = 'IS_IFKEYS'.
+        ls_parameter-kind = cl_abap_objectdescr=>exporting.
+        GET REFERENCE OF is_ifkeys INTO ls_parameter-value.
+        INSERT ls_parameter INTO TABLE lt_parameters.
+
+        ls_parameter-name = 'RV_SUCCESS'.
+        ls_parameter-kind = cl_abap_objectdescr=>returning.
+        GET REFERENCE OF rv_success INTO ls_parameter-value.
+        INSERT ls_parameter INTO TABLE lt_parameters.
+
+        CALL METHOD mr_abapgit_util->(lv_meth)
+          PARAMETER-TABLE
+          lt_parameters.
```
There's no need to use `PARAMETER-TABLE` in dynamic method call. Supplying parameters explicitly will make maintenance a bit easier and the code cleaner. What do you think?

This is e.g. how it's usually done in other places.
https://github.com/abapGit/abapGit/blob/fe363f7b0a9681302cc3ec320211c563bdb4f7b4/src/objects/zcl_abapgit_object_ddls.clas.abap#L141-L145

D047539 replied:
Coding was adjusted as requested.

D047539 commented:
> thanks, can you help opening a PR to https://github.com/abapGit-tests/AIFC with a simple example of the AIFC object
> 
> we keep examples for testing and future reference

Please see: https://github.com/abapGit-tests/AIFC/pull/1, hope the AIFC test content will help

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+
+    INTERFACES zif_abapgit_object.
+
+    TYPES:
+      BEGIN OF ty_content_s,
+        tabname TYPE tabname,
+      END OF ty_content_s.
+    TYPES:
+      ty_content_t TYPE STANDARD TABLE OF ty_content_s WITH NON-UNIQUE DEFAULT KEY .
```
These types could be moved to `PRIVATE SECTION`

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+        ifver  TYPE c LENGTH 5,
+      END OF ty_aif_key_s .
+    TYPES:
+******************************************************************
```
Suggest to remove these ***-lines

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+    DATA lr_table TYPE REF TO data.
+    FIELD-SYMBOLS <lt_table> TYPE STANDARD TABLE.
+    FIELD-SYMBOLS <ls_table> TYPE any.
+    FIELD-SYMBOLS <ls_entry> TYPE any.
+    FIELD-SYMBOLS <lv_spras> TYPE any.
```
These two field-symbols are not used. I'd suggest to remove them.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+
+  METHOD zif_abapgit_object~get_metadata.
+
+    CLEAR rs_metadata.
```
No need to clear `RETURNING` parameter.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+
+  METHOD get_content_compress.
+    DATA: lr_dyn_call_error TYPE REF TO cx_sy_dyn_call_error.
+    DATA: lr_root TYPE REF TO cx_root.
```
Exception objects are usually prefixed with lx.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,536 @@
+  METHOD get_content_compress.
+    DATA: lr_dyn_call_error TYPE REF TO cx_sy_dyn_call_error.
+    DATA: lr_root TYPE REF TO cx_root.
+    DATA: lr_log TYPE REF TO object.
```
Objects are usually prefixed with lo.

christianguenter2 reviewed (commented):
added few more comments...

D047539 commented:
> 

thanks a lot for your feedback, I changed the coding as requested

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+    super->constructor( is_item     = is_item
+                        iv_language = iv_language ).
+
+    CLEAR ms_icd_data_key.
```
```suggestion
```
No need to clear attributes in `CONSTRUCTOR`

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+            rv_success = rv_success.
+
+      CATCH cx_sy_dyn_call_error INTO lx_dyn_call_error.
+        zcx_abapgit_exception=>raise( iv_text = TEXT-003
```
Text symbols are not supported by abapGit standalone version. Therefore we use hard coded texts instead.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+          TRY.
+              lv_tablename = cl_abap_dyn_prg=>check_table_name_str( val = lr_content->tabname
+                                                                    packages = '' ).
+            CATCH cx_abap_not_a_table.
+              RAISE EXCEPTION TYPE zcx_abapgit_exception.
```
suggest to use `zcx_abapgit_exception=>raise` instead of `RAISE EXCEPTION TYPE zcx_abapgit_exception.` or supply ix_previous. Because the latter will only show a generic 'An exception was raised' to the user which is usually not very helpful.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+    DATA lr_ifdata TYPE REF TO ty_table_data_s.
+    FIELD-SYMBOLS <lt_table> TYPE ANY TABLE.
+
+    IF zif_abapgit_object~exists( ) = abap_false.
+      RETURN.
+    ENDIF.
```
```suggestion
```
There's no need for existence check here. It's already checked by abapGit serializer logic before.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+        io_xml->add( iv_name = `Content_table`
+                     ig_data = lt_content ).
+
+      CATCH cx_root INTO lx_root.
+    ENDTRY.
```
Looks suspicious. Above raised exceptions are catched silently here. Suggest to propagate them.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,509 @@
+          RECEIVING
+            rr_abapgit_aifc_util = mo_abapgit_util.
+
+      CATCH cx_sy_dyn_call_error INTO lx_exc_ref.
+    ENDTRY.
```
When I try to test the demo repo in my system I get this shotdump because it doesn't support AIFC objects.
![grafik](https://user-images.githubusercontent.com/17437789/146135812-f58a8de7-c687-4801-8ab2-3d13b59dd7ba.png)

We can fix it like this:
```suggestion
      CATCH cx_sy_dyn_call_error INTO lx_exc_ref.
          zcx_abapgit_exception=>raise( 'AIFC not supported' ).
    ENDTRY.
```
and add `raising zcx_abapgit_exception` to `CONSTRUCTOR` signature.

christianguenter2 reviewed (changes requested):
@D047539 Added a few more comments. Can you check them?

D047539 commented:
> 

I checked the coding as requested and adapted the exception handling and remove the existence check and text symbols

christianguenter2 commented:
@D047539 what are the prerequisites to install the test repo?
Cannot install it in my S/4HANA 2020 system.
![grafik](https://user-images.githubusercontent.com/17437789/146174156-21b40b6d-0e31-4972-9042-67c1856293a0.png)

D047539 commented:
> @D047539 what are the prerequisites to install the test repo? Cannot install it in my S/4HANA 2020 system. ![grafik](https://user-images.githubusercontent.com/17437789/146174156-21b40b6d-0e31-4972-9042-67c1856293a0.png)

The two following notes are required as prerequisites for using ABAP GIT functionality
- 3106807-AIF contentent delivery adoption for ABAP GIT (https://launchpad.support.sap.com/#/notes/3106807)
- 3117150-AIF adoption for abapGit (https://launchpad.support.sap.com/#/notes/3117150)

christianguenter2 commented:
@D047539 
I cannot install these notes in my system
![grafik](https://user-images.githubusercontent.com/17437789/146176659-e55f3486-2c13-4567-8499-97d745e0bae0.png)

We're on SAP_BASIS 755 SP00

D047539 commented:
> @D047539 I cannot install these notes in my system ![grafik](https://user-images.githubusercontent.com/17437789/146176659-e55f3486-2c13-4567-8499-97d745e0bae0.png)
> 
> We're on SAP_BASIS 755 SP00

Due to technical requirements, we cannot release the Notes for earlier SPs.
Is there no possibility to install the required SP into test system?

christianguenter2 commented:
What is the required SP? Can I see this information in the note?

D047539 commented:
> What is the required SP? Can I see this information in the notes?

The required SP is SP01. You will see the information in the technical details of the corrections in SNOTE. There you can find the part Validities with the information of the Validity Range.

christianguenter2 commented:
I also have a SP01 system at my hand where I implemented the notes. While importing the demo project I get now this error:
![grafik](https://user-images.githubusercontent.com/17437789/146188251-77ee5c54-6b7a-413e-a529-afabffab7496.png)
![grafik](https://user-images.githubusercontent.com/17437789/146188259-e422a99e-e5c5-43bc-baeb-040b2ab891c4.png)
![grafik](https://user-images.githubusercontent.com/17437789/146188542-f4a1537a-282d-4c21-b466-3fe3f639d137.png)
![grafik](https://user-images.githubusercontent.com/17437789/146188743-7b6948c4-39fd-4f3b-aec6-840736868af7.png)

mbtools commented:
The XML contains the client number:
https://github.com/abapGit-tests/AIFC/blob/main/src/abapgit_aifc_test_01z_git%20flbooking%20001.aifc.xml#L7

I think this will lead to diffs since users work with different clients. Suggest removing `mandt` from the XML i.e. clear it before adding the data to `ii_xml`.

D047539 commented:
I don't think that the field mandt is causing the issue. I suspect a setting in transport management does not allow this combination of TADIR object and package. 
I will upload a new variant for the test content.

mbtools commented:
My comments were not related to the issue Christian mentioned. These are things that matter in general.

christianguenter2 commented:
I removed the old object from the test repo. Now I was able to pull the object. 
![grafik](https://user-images.githubusercontent.com/17437789/146321382-3964b9c1-5b46-4ac1-9eb0-dca6ae705afc.png)


As Marc already noted there's a diff regarding the client field.
![grafik](https://user-images.githubusercontent.com/17437789/146321440-c3815e06-10cc-41f7-8cb0-25ed98834006.png)
![grafik](https://user-images.githubusercontent.com/17437789/146320388-7e4eccdd-91cf-4ed7-9574-24e1a3aa69f8.png)

D047539 commented:
> I removed the old object from the test repo. Now I was able to pull the object. ![grafik](https://user-images.githubusercontent.com/17437789/146321382-3964b9c1-5b46-4ac1-9eb0-dca6ae705afc.png)
> 
> As Marc already noted there's a diff regarding the client field. ![grafik](https://user-images.githubusercontent.com/17437789/146321440-c3815e06-10cc-41f7-8cb0-25ed98834006.png) ![grafik](https://user-images.githubusercontent.com/17437789/146320388-7e4eccdd-91cf-4ed7-9574-24e1a3aa69f8.png)

I removed the client field from the xml file and replaced the exception call as requested. Furthermore I send new AIFC test content into the AIFC Test repository.

mbtools commented:
Looks good. There's a conflict on https://github.com/abapGit-tests/AIFC/pull/3. Please check

D047539 commented:
> Looks good. There's a conflict on [abapGit-tests/AIFC#3](https://github.com/abapGit-tests/AIFC/pull/3). Please check

Thanks. Conflict is solved.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,547 @@
+    ls_icd_data_key-ifname = ms_icd_data_key-ifname.
+    ls_icd_data_key-ifver2 = ms_icd_data_key-ifver2.
+
+    CLEAR rv_user.
```
No need to clear returning parameter.

christianguenter2 commented on src/objects/zcl_abapgit_object_aifc.clas.abap:
```diff
@@ -0,0 +1,547 @@
+
+
+  METHOD clear_client.
+    DATA lx_root TYPE REF TO cx_root.
+    DATA lr_structdescr TYPE REF TO cl_abap_structdescr.
+    DATA lr_tabledescr TYPE REF TO cl_abap_tabledescr.
+    DATA lr_typedescr TYPE REF TO cl_abap_typedescr.
+    DATA lv_type TYPE string.
+    DATA lv_name TYPE c LENGTH 30.
+    DATA lt_components TYPE cl_abap_structdescr=>component_table.
+    FIELD-SYMBOLS: <ls_components> LIKE LINE OF lt_components.
+    FIELD-SYMBOLS <ls_table> TYPE any.
+    FIELD-SYMBOLS <lv_value> TYPE any.
+
+    TRY.
+        lr_tabledescr ?= cl_abap_typedescr=>describe_by_data( p_data = ct_data  ).
+        lr_structdescr ?= lr_tabledescr->get_table_line_type( ).
+
+        lt_components = lr_structdescr->get_components( ).
+        LOOP AT lt_components ASSIGNING <ls_components>.
+          CHECK <ls_components>-type IS ASSIGNED.
+          lr_typedescr ?= <ls_components>-type.
+          lv_type = lr_typedescr->get_relative_name( ).
+          IF lv_type = 'MANDT'.
+            lv_name = <ls_components>-name.
+            EXIT.
+          ENDIF.
+        ENDLOOP.
+
+        LOOP AT ct_data ASSIGNING <ls_table>.
+          ASSIGN COMPONENT lv_name OF STRUCTURE <ls_table> TO <lv_value>.
+          CHECK <lv_value> IS ASSIGNED.
+          CLEAR <lv_value>.
+        ENDLOOP.
+      CATCH cx_root INTO lx_root.
+        zcx_abapgit_exception=>raise_with_text( lx_root ).
+    ENDTRY.
```
Suggest to simplify logic like this:

```suggestion
    DATA:
      BEGIN OF ls_data_to_clear,
        mandt TYPE sy-mandt,
      END OF ls_data_to_clear.

    FIELD-SYMBOLS: 
      <data> TYPE any.

    LOOP AT ct_data ASSIGNING <data>.
      MOVE-CORRESPONDING ls_data_to_clear TO <data>.
    ENDLOOP.

```

christianguenter2 reviewed (changes requested):
Two last comments from my side regarding the code.

Format looks good now, no more diffs.
![grafik](https://user-images.githubusercontent.com/17437789/146506953-3b5dfb4a-00dd-4779-86f5-d573677f993f.png)

D047539 commented:
> Two last comments from my side regarding the code.
> 
> Format looks good now, no more diffs. ![grafik](https://user-images.githubusercontent.com/17437789/146506953-3b5dfb4a-00dd-4779-86f5-d573677f993f.png)

I updated the coding as requested

christianguenter2 reviewed (approved):
👍

larshp commented:
welcome to abapGit as a contributor 🎉
